### PR TITLE
Emoji shortcodes 

### PR DESCRIPTION
**Pull request by TheDrawingCoder-Gamer**
Adds emoji shortcode support. Also manages mangling names.

NOTE: I don't write in python, so any help in making in not horrible code would be appreciated

The WIP is because there aren't as many short codes as would be necessary. 

Solves #1052

### CODE REVIEW

File: scripts/emoji_codegen.py [hunk 1]
```diff
@@ -68,16 +73,34 @@ def generate_qml_list(**kwargs):
...
 
-        if qualification == 'component':
-            continue
 
         char, name = re.match(r'^(\S+) E\d+\.\d+ (.*)$', charAndName).groups()
-
+        # drop "face" part
+        
+        if name in shortcodeDict: 
+            name = shortcodeDict[name]
```
[the remark this answers is not in the document]
Comment (TheDrawingCoder-Gamer) [reply scripts/emoji_codegen.py:84]: Wouldn't that mean it's populated with two emojis in the menu?

**Review by TheDrawingCoder-Gamer — COMMENTED** [review]
Flatpak is broken - doesn't seem related to what I am doing

File: scripts/emoji_codegen.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheDrawingCoder-Gamer) [reply scripts/emoji_codegen.py]: wdym additional entry?

[the remark this answers is not in the document]
Comment (TheDrawingCoder-Gamer) [reply scripts/emoji_codegen.py]: I have no CPP experience lol, i'll try my best

**Comment by deepbluev7** [comment]
Yeah, the flatpak is broken because a PPA got removed.

**Comment by TheDrawingCoder-Gamer** [comment]
How should I handle emojis w/ different skin tones?

**Comment by TheDrawingCoder-Gamer** [comment]
Even though I made Qt::Tooltip return unicode name the tooltip still shows the short name

**Comment by deepbluev7** [comment]
No idea about the skintones, but wouldn't it make more sense to have the unicodename be the name from the unicode files and the shortnames the one you added? Or am I reading the code wrong? :D

**Comment by TheDrawingCoder-Gamer** [comment]
The shortcodes are the ones I added, unless I did something reallt wrong

**Comment by TheDrawingCoder-Gamer** [comment]
I am unsure of why it's not displaying and I don't know enough C++ to fully understand this repo.

**Comment by TheDrawingCoder-Gamer** [comment]
No explicit references to the shortname role :frowning:

**Comment by TheDrawingCoder-Gamer** [comment]
Swapping SearchRole doesn't work either

**Comment by TheDrawingCoder-Gamer** [comment]
I don't know enough about this project to fully fix this - Please help me with this :sob:

**Comment by TheDrawingCoder-Gamer** [comment]
Ok, the reason it was showing that was the Qml was asking for model.shortName, so I changed it to model.unicodeName

**Comment by TheDrawingCoder-Gamer** [comment]
What python version should I use? is 3.8 an OK requirement?

**Comment by TheDrawingCoder-Gamer** [comment]
I would like to add something like discords' skintone selector but am not very fluent in qml, any idea how I would do this?

File: scripts/emoji_codegen.py [hunk 2]
```diff
@@ -68,16 +105,67 @@ def generate_qml_list(**kwargs):
...
-        categories[current_category].append(Emoji(code, name))
+        shortname = name
+        
+        # discard skin tone variants for sanity
+        # __contains__ is so stupid i hate prototype languages
+        if name.__contains__("skin tone") and qualification != 'component': 
+            continue
+        if qualification == 'component' and not name.__contains__("skin tone"): 
+            continue
+        #TODO: Handle skintone modifiers in a sane way
```
Comment (deepbluev7) [inline scripts/emoji_codegen.py:121]: Until we handle skin tones differently, can we keep them in the picker? It would be annoying to not be able to send those variants anymore...

File: scripts/emoji_codegen.py [hunk 3]
```diff
@@ -68,16 +105,67 @@ def generate_qml_list(**kwargs):
         code, qualification, charAndName = segments
 
         # skip unqualified versions of same unicode
-        if qualification == 'unqualified':
-            continue
-
-        if qualification == 'component':
+        if qualification != 'fully-qualified' and qualification != 'component' :
```
Comment (deepbluev7) [inline scripts/emoji_codegen.py:108]: Why did you change this condition?

File: scripts/emoji_codegen.py [hunk 4]
```diff
@@ -68,16 +105,67 @@ def generate_qml_list(**kwargs):
...
+                shortname = shortname[:-5]
+            if shortname.endswith(' button'): 
+                shortname = shortname[:-7] 
+            if shortname.endswith(' banknote'): 
+                shortname = shortname[:-9]
+            keycapmtch = re.match(r'^keycap: (.+)$', shortname)
+            if keycapmtch: 
+                keycapthing, = keycapmtch.groups()
+                type(keycapthing)
+                num_name = humanize_keypad(keycapthing)
```
Comment (deepbluev7) [inline scripts/emoji_codegen.py:148]: Why would you want to type out the number instead of just using your number key?

File: src/emoji/EmojiModel.cpp [hunk 5]
```diff
@@ -58,10 +59,12 @@ EmojiModel::data(const QModelIndex &index, int role) const
             return Provider::emoji[index.row()].unicode;
 
         case Qt::ToolTipRole:
```
Comment (deepbluev7) [inline src/emoji/EmojiModel.cpp:61]: Can you make the tooltip include both shortname and unicode, if they are different and not empty?

File: scripts/emoji_codegen.py (hunk 3 above)
Comment (TheDrawingCoder-Gamer) [reply scripts/emoji_codegen.py:108]: To allow skin tone modifiers to be displayed

File: src/emoji/EmojiModel.cpp (hunk 5 above)
Comment (deepbluev7) [reply src/emoji/EmojiModel.cpp:61]:
```cpp
case Qt::ToolTipRole:
{
    if (Provider::emoji[index.row()].shortcode.isEmpty())
        return Provider::emoji[index.row()].unicode;
    else
        return Provider::emoji[index.row()].shortcode + ", " + Provider::emoji[index.row()].unicode:
}
```

File: scripts/emoji_codegen.py [hunk 6]
```diff
@@ -68,16 +74,65 @@ def generate_qml_list(**kwargs):
...
             continue
+        
 
         char, name = re.match(r'^(\S+) E\d+\.\d+ (.*)$', charAndName).groups()
-
-        categories[current_category].append(Emoji(code, name))
+        shortname = name
+        # until skin tone is handled, keep them around
+        # discard skin tone variants for sanity
+        # __contains__ is so stupid i hate prototype languages
```
Comment (deepbluev7) [inline scripts/emoji_codegen.py:85]: You wouldn't call `__contains__` directly, you would do `if "skin tone" in name:`

**Comment by deepbluev7** [comment]
Thank you!
